ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java

```diff
@@ -281,4 +262,121 @@ private void unloadErrorsFromCache(final ImmutableMap<K, ComposableFuture<V>> in
+    }
+
+    public Builder<K, V> withLoader(final CacheLoader<K, V> loader) {
+      this.loader = loader;
+      return this;
+    }
+
+    public Builder<K, V> withLoadTimeout(final long loadTimeout, final TimeUnit loadTimeoutUnit) {
+      this.loadTimeout = loadTimeout;
+      this.loadTimeoutUnit = loadTimeoutUnit;
+      return this;
+    }
+
+    public Builder<K, V> failOnMissingEntries() {
+      this.failOnMissingEntries = true;
+      return this;
+    }
+
+    public Builder<K, V> refreshAfterWrite(long duration, TimeUnit unit) {
+      this.refreshAfterWriteDuration = duration;
+      this.refreshAfterWriteUnit = unit;
+      return this;
+    }
+
+    public Builder<K, V> withRefreshThreadCount(int refreshThreadCount) {
+      this.refreshThreadCount = refreshThreadCount;
+      return this;
+    }
+
+    public LocalAsyncCache<K, V> build() {
+      if (loader == null) {
+        return new LocalAsyncCache<>(maximumSize, ttl, unit, metricFactory, cacheName);
+      }
+      return new LocalAsyncCache<>(maximumSize, ttl, unit, loader, metricFactory, cacheName, failOnMissingEntries,
+              loadTimeout, loadTimeoutUnit, refreshAfterWriteDuration, refreshAfterWriteUnit, refreshThreadCount);
+    }
+  }
+
+  private class InternalCacheLoader extends com.google.common.cache.CacheLoader<K, ComposableFuture<V>> {
+
+    private final CacheLoader<K, V> loader;
+    private final String cacheName;
+    private final boolean failOnMissingEntries;
+    private final long loadTimeout;
+    private final TimeUnit loadTimeoutUnit;
+
+    InternalCacheLoader(final CacheLoader<K, V> loader, final String cacheName, final boolean failOnMissingEntries,
+                        final long loadTimeout, final TimeUnit loadTimeoutUnit) {
+      this.loader = loader;
+      this.cacheName = cacheName;
+      this.failOnMissingEntries = failOnMissingEntries;
+      this.loadTimeout = loadTimeout;
+      this.loadTimeoutUnit = loadTimeoutUnit;
+    }
+
+    @Override
+    public ComposableFuture<V> load(@Nonnull final K key) {
+      return loader.load(cacheName, key).withTimeout(loadTimeout, loadTimeoutUnit).materialize();
+    }
+
+    @Override
+    public Map<K, ComposableFuture<V>> loadAll(final Iterable<? extends K> keys) {
+      final ComposableFuture<Map<K, V>> loaded = loader.load(cacheName, keys).withTimeout(loadTimeout, loadTimeoutUnit).materialize();
+      final Map<K, ComposableFuture<V>> result = new HashMap<>();
+      for (final K key : keys) {
+        result.put(key, loaded.flatMap(extractLoaderResultEntry(key)));
+      }
+      return result;
+    }
+
+    @Override
+    public ListenableFuture<ComposableFuture<V>> reload(final K key, final ComposableFuture<V> oldValue) {
+      ListenableFutureTask<ComposableFuture<V>> task = ListenableFutureTask.create(() -> {
+        ComposableFuture<V> loadFuture = loader.load(cacheName, key).recoverWith(e -> oldValue);
+        V value = loadFuture.get(loadTimeout, loadTimeoutUnit);
```

Comment:
You can implement reload this way:
      final SettableFuture<ComposableFuture<V>> future = SettableFuture.create();

      loader.load(cacheName, key).materialize().andThen(t -> {
        if (t.isSuccess()) {
          future.set(ComposableFutures.fromValue(t.getValue()));
        } else {
          future.setException(t.getError());
        }
      });

      return future;

Reply:
this is just perfect! thanks!